### DISCUSSION TITLE
Slow UDP over 100Gb NIC

### DISCUSSION THREAD

**Discussion by VolunPotest** [esnet/iperf#1274] [Q&A]
Hi! 
We have in lab 100Gb NIC, and we are trying to measure TCP and UDP throughput.

The max we can get for TCP ~ 50 GBits/s
The max we can get with UDP ~ 12 Gbit/s 

My question is: 
Why is UDP throughput so low (comparing to TCP). I heard a lot that UDP should be much faster than TCP, but in real example we have the opposite. 
If there is any explanation about that? Is it connected with Linux Kernel itself or there is a problem with iperf3? 

We usually use basic iperf3 commands like:
iperf3 -c [server_ip] -t 30 -V -u -b 0

* Version of iperf3: 3.7
* Operating system (and distribution, if any): Ubuntu 20.04 LTS

Thank you!

**Comment by davidBar-On** [comment]
This may be because of the packets sizes - the default for TCP and UPD are different.  Use the `-d` option to see the value used - the `len` value in `get_parameters` JSON.

To compare apples to apples, the same packet size should be used for both, using the `-l` option.

**Comment by VolunPotest** [comment]
Hi!
Thank you for your answer.

Packet size for TCP is 128K (by default) and for UDP is 1448 bytes. (MTU 1500), but as far as I understand, NIC will split this big TCP packet for small ~1500 bytes chunks (because packet size couldn't be more than MTU size). 

Okay. I've tried to increase packet size for UDP with -l option and maximum is 64K, but I can see that is not UDP packets anymore, it's IP fragments packets, but throughput is ~15 Gbits/s now. (Not huge difference)

I also tried to make an opposite. I've made the packet size for TCP and UDP equal to 1400:
- For TCP results is ~6Gbits./s
- For UDP results is ~4Gbit/s.

Still, UDP is lower than TCP

**Comment by lomaxfrog** [comment]
Hi,

If you are using Mellanox ConnectX5, it might be that you are running into
the issue that only the first ethernet frame of your packet get properly
dispatched into the receiving queues of the NIC (load balancing the frames
on different IRQ/cores) and that the remaining frames all go to queue 0,
i.e. to the same IRQ/core. You can verify that if you see a single core on
the receive side being max'ed up, or looking at interrupt counts and see if
a queue is overwhelmed.

Eric

**Comment by VolunPotest** [comment]
Thank you for the answer, @lomaxfrog.
I don't remember the model name, but I think it's intel network card. 

I didn't really understand what you're trying to tell me about IRQ/core. 
Could you please recommend me an article/guidance or something like that to understand this issue more properly?

Thank you

**Comment by swlars** [comment]
Hi,

A bigger socket buffer size might help as well, using the -w option.

**Comment by kgersen** [comment]
check this: https://github.com/lucas-clemente/quic-go/wiki/UDP-Receive-Buffer-Size
